src/lib/utils/api.ts

```diff
@@ -12,6 +13,47 @@ export function isRedirectCode(
   return status >= 300 && status <= 308;
 }
 
+/**
+ * Handle what comes back from the API and return either data or errors.
+ *
+ * Two generic types are passed through:
+ *
+ * - T is data from the API
+ * - E is an error coming back from the API
+ *
+ * @param resp The fetch response from the API. If this is missing, fetch
+ * threw an error and we should send a 500 to the user because the API is
+ * probably down.
+ */
+export async function getApiResponse<T, E>(
+  resp?: Response | void,
+): Promise<APIResponse<T, E>> {
+  const response: APIResponse<T, E> = {};
+
+  if (!resp) {
+    response.error = {
+      status: 500,
+      message: "API error",
+    };
+
+    return response;
+  }
+
+  if (isErrorCode(resp.status)) {
+    response.error = {
+      status: resp.status,
+      message: resp.statusText,
+      errors: await resp.json(),
+    };
+
+    return response;
+  }
+
+  // everything worked
+  response.data = await resp.json();
```

Comment:
We should guard the call to `json`, since it can fail with a `SyntaxError` if the response body is malformed JSON (this can happen if the API returns its own 500 error page) or if the body is empty in a 204 No Content response.

```suggestion
  try {
    response.data = await resp.json();
  } catch (e) {
    switch (e.name) {
      case "SyntaxError": // provide more specific error
        response.error = {
          status: 500,
          message: "The API returned invalid JSON"
        }
        break;
      default: // catch-all handling
        response.error = {
          status: 500,
          message: String(e), 
        }
    }
  }
```